File: src/main/java/tech/flandia_yingm/auto_fgo/device/Device.java
package tech.flandia_yingm.auto_fgo.device;

import lombok.val;
import org.slf4j.LoggerFactory;
import tech.flandia_yingm.auto_fgo.img.Images;
import tech.flandia_yingm.auto_fgo.img.Point;
import tech.flandia_yingm.auto_fgo.script.Script;
import tech.flandia_yingm.auto_fgo.img.Template;

import java.awt.image.BufferedImage;
import java.util.Arrays;
import java.util.List;

public interface Device {

    void tap(Point point);

    void swipe(Point start, Point end, long duration);

    void insert(String text);

    BufferedImage capture();


    default boolean match(Template template) {
        val log = LoggerFactory.getLogger(this.getClass());

        log.debug("{} - Matching template: {}", this, template);
        val similarity = Images.matchSameTemplate(capture(), template.getImage());
        val match = similarity > template.getThreshold();
        if (match) {
            log.debug("{} - Matched template: {}, similarity: {}", this, template, similarity);
        } else {
            log.debug("{} - Didn't matched template: {}, similarity: {}", this, template, similarity);
        }
        return match;
    }

    default boolean match(Template... templates) {
        return match(Arrays.asList(templates));
    }

    default boolean match(List<Template> templateList) {
        return templateList.stream()
                           .map(this::match)
                           .reduce(false, (a, b) -> a || b);
    }

    default Point find(Template template) {
        val log = LoggerFactory.getLogger(this.getClass());

        log.debug("{} - Finding template: {}", this, template);
        val similarityPoint = Images.matchTemplate(capture(), template.getImage());
        if (similarityPoint.getWeight() > template.getThreshold()) {
            return similarityPoint;
        } else {
            return Point.getEmpty();
        }
    }

    default void delay(long ms) {
        val log = LoggerFactory.getLogger(this.getClass());
        try {
            log.debug("{} - Delaying {} ms", this, ms);
            Thread.sleep(ms);
            log.debug("{} - Delayed {} ms", this, ms);
        } catch (InterruptedException e) {
            //Ignored: thread interrupted
        }
    }


    default void run(Script script) {
        script.runScript(this);
    }

}

File: src/main/java/tech/flandia_yingm/auto_fgo/device/Device.kt
package tech.flandia_yingm.auto_fgo.device

import org.slf4j.Logger
import org.slf4j.LoggerFactory
import tech.flandia_yingm.auto_fgo.img.Images.findTemplate
import tech.flandia_yingm.auto_fgo.img.Images.matchTemplate
import tech.flandia_yingm.auto_fgo.img.Point
import tech.flandia_yingm.auto_fgo.img.Point.Companion.EMPTY
import tech.flandia_yingm.auto_fgo.img.Template
import java.awt.image.BufferedImage

interface Device {

    private val log: Logger
        get() = LoggerFactory.getLogger(this.javaClass)


    fun tap(point: Point)

    fun swipe(start: Point, end: Point, duration: Long)

    fun insert(text: String)

    fun capture(): BufferedImage


    fun delay(ms: Long) {
        val log = LoggerFactory.getLogger(this.javaClass)
        try {
            log.debug("{} - Delaying {} ms", this, ms)
            Thread.sleep(ms)
            log.debug("{} - Delayed {} ms", this, ms)
        } catch (e: InterruptedException) {
            //Ignored: thread interrupted
        }
    }


    fun isMatching(template: Template): Boolean {
        log.debug("{} - Matching template: {}", this, template)
        val similarity = matchTemplate(capture(), template.image)
        val matching = similarity > template.threshold
        if (matching) {
            log.debug("{} - Matched template: {}, similarity: {}", this, template, similarity)
        } else {
            log.debug("{} - Didn't match template: {}, similarity: {}", this, template, similarity)
        }
        return matching
    }

    fun isMatching(vararg templates: Template): Template {
        return isMatching(listOf(*templates))
    }

    fun isMatching(templates: List<Template>): Template {
        log.debug("{} - Matching templates {}", this, templates)
        val image = capture()
        for (template in templates) {
            val similarity = matchTemplate(image, template.image)
            if (similarity > template.threshold) {
                log.debug("{} - Matched template {}, similarity: {}", this, template, similarity)
                return template
            } else {
                log.debug("{} - Didn't match template {}, similarity: {}", this, template, similarity)
            }
        }
        return Template.EMPTY_TEMPLATE
    }

    fun waitMatching(template: Template) {
        while (!Thread.interrupted()) {
            val matching = isMatching(template)
            if (matching) {
                return
            }
        }
    }

    fun waitMatching(vararg templates: Template): Template {
        return waitMatching(listOf(*templates))
    }

    fun waitMatching(templates: List<Template>): Template {
        while (!Thread.interrupted()) {
            val matching = isMatching(templates)
            if (matching != Template.EMPTY_TEMPLATE) {
                return matching
            }
        }
        throw InterruptedException()
    }


    fun findMatching(template: Template): Point {
        log.debug("{} - Finding the template: {}", this, template)
        val sp = findTemplate(capture(), template.image)
        log.debug("{} - Found the template, point: {}", this, sp)
        return if (sp.weight > template.threshold) sp else EMPTY
    }

}